### ISSUE TITLE
P1208R6 <source_location>

### ISSUE THREAD

**Issue by StephanTLavavej** [microsoft/STL#54]
[P1208R6](https://wg21.link/P1208R6) `<source_location>`

LWG-3396 Clarify point of reference for `source_location::current()` (DE 169)

This needs compiler support.

The IDE's list of extensionless headers has been updated for VS 2019 16.5.

Feature-test macro as of WG21-N4842:
`#define __cpp_lib_source_location 201907L`

**Comment by StephanTLavavej** [comment]
Microsoft-internal MSVC-PR-229114 implemented intrinsics for VS 2019 16.6 Preview 2:

```cpp
const char* __builtin_FILE();
const char* __builtin_FUNCTION();
int __builtin_LINE();
int __builtin_COLUMN();
```

**Comment by SuperWig** [comment]
@StephanTLavavej Is this not also blocked on `consteval`?

**Comment by StephanTLavavej** [comment]
Aha, `consteval` didn't exist when I first looked at `source_location` years ago, but it's there in the paper now. I'm unsure as to how much of a blocker this is; thanks for bringing it up!

**Comment by SuperWig** [comment]
>I'm unsure as to how much of a blocker this is

Isn't not being able to compile (including tests, which from what I've read makes Billy sad) a pretty big blocker? 

Having it be `constexpr` instead doesn't seem like a correct thing to do as that's not what the paper says 😕.

**Comment by StephanTLavavej** [comment]
Thanks for the investigation; we'll need to talk to the compiler devs again.

**Comment by SuperWig** [comment]
One question I forgot to ask was, where is the value of the column supposed to be
```
source_location::current();
                 ^
source_location::current();
^
```
Because I got the first but kinda expected the latter.

**Comment by AdamBucior** [comment]
> One question I forgot to ask was, where is the value of the column supposed to be?
> 
> ```
> source_location::current();
>                  ^
> source_location::current();
> ^
> ```
> 
> Because I got the first but kinda expected the latter.

Paper describes the column value as implementation-defined, so I suppose it's OK. Also both of these values seem fine to me as I can't imagine a situation in which it'd matter.

**Comment by SuperWig** [comment]
Yeah ,I saw that. I guess I was more or less asking if it was intentional for it to be that position and/or if one was preferred, rather than asking which is "correct".

**Comment by StephanTLavavej** [comment]
VS 2019 16.6 Preview 2 shipped on March 26, 2020.

**Comment by BurningEnlightenment** [comment]
May I ask, whether it is planned to enhance `__builtin_FUNCTION()` in a future release to include function template parameters, etc.? It feels a bit strange that the new C++20 solution produces inferior output than the old `__FUNCSIG__` macro. Would a devcom issue regarding this help? (I'm unsure, because opening an issue about something that technically isn't available might be a reason to close it)

**Comment by StephanTLavavej** [comment]
@BurningEnlightenment Yes, I recommend submitting a DevCom suggestion. Just make it clear that you're requesting a change to the `__builtin_FUNCTION()` intrinsic that powers C++20 `source_location`; you can even link to #664 and mention that it's been merged for VS 2019 16.10 Preview 2, and that the STL maintainers have asked you to submit a suggestion :smile_cat:

**Comment by Dwarfius** [comment]
Thank you for the implementation of <source_location>. Per above discussion, I created a [DevCom feedback ticket](https://developercommunity.visualstudio.com/t/Improvement-Expand-output-of-__builtin/1491652) to expand the output of __builtin_FUNCTION with more details about the context. I'm leaving it here in case someone else is looking for information on the situation/progress of it.

Happy to nuke the comment if it's not the right place for this information.
